test/mpi/dtpools/src/dtpools_internal.c

```diff
@@ -0,0 +1,2043 @@
+    } else if (basic_type == MPI_FLOAT_INT) {
+        DTPI_OBJ_CHECK_COMP_BUF_AND_JUMP(dtp_float_int, float, int, ptrs.float_int_ptr);
+    } else if (basic_type == MPI_DOUBLE_INT) {
+        DTPI_OBJ_CHECK_COMP_BUF_AND_JUMP(dtp_double_int, double, int, ptrs.double_int_ptr);
+    } else if (basic_type == MPI_LONG_INT) {
+        DTPI_OBJ_CHECK_COMP_BUF_AND_JUMP(dtp_long_int, long, int, ptrs.long_int_ptr);
+    } else if (basic_type == MPI_2INT) {
+        DTPI_OBJ_CHECK_COMP_BUF_AND_JUMP(dtp_2int, int, int, ptrs.int_int_ptr);
+    } else if (basic_type == MPI_SHORT_INT) {
+        DTPI_OBJ_CHECK_COMP_BUF_AND_JUMP(dtp_short_int, short, int, ptrs.short_int_ptr);
+    } else if (basic_type == MPI_LONG_DOUBLE_INT) {
+        DTPI_OBJ_CHECK_COMP_BUF_AND_JUMP(dtp_long_double_int, long double, int,
+                                         ptrs.long_double_int_ptr);
+    }
+
+  fn_exit:
+    return err;
+
+  fn_fail:
+    err = DTP_ERR_OTHER;
+    goto fn_exit;
+}
+
+/* --------------------------------------------------------- */
+/* Datatype Pool Object Create Functions                     */
+/* --------------------------------------------------------- */
+
+int DTPI_Struct_create(DTPI_Bound_e bound, struct DTPI_Par *par, DTP_t dtp)
+{
+    int i;
+    int err = DTP_SUCCESS;
+    int num_types = 0;
+    int obj_idx;
+    int *basic_type_counts;
+    MPI_Aint *basic_type_sizes = NULL;
+    MPI_Aint lb, extent;
+    MPI_Aint displs = 0;
+    MPI_Aint *basic_type_displs = NULL;
+    MPI_Datatype *basic_types;
+    void *buf = NULL;
+    DTPI_t *dtpi = NULL;
+    char type_name[TYPE_NAME_MAXLEN] = { 0 };
+
+    /* get object index */
+    obj_idx = par->user.obj_idx;
+
+    num_types = dtp->DTP_type_signature.DTP_pool_struct.DTP_num_types;
+    DTPI_OBJ_ALLOC_OR_FAIL(basic_type_sizes, sizeof(*basic_type_sizes) * num_types);
+    DTPI_OBJ_ALLOC_OR_FAIL(basic_type_displs, sizeof(*basic_type_displs) * num_types);
+
+    /* compute struct size */
+    basic_types = dtp->DTP_type_signature.DTP_pool_struct.DTP_basic_type;
+    basic_type_counts = dtp->DTP_type_signature.DTP_pool_struct.DTP_basic_type_count;
+    for (i = 0; i < num_types; i++) {
+        err = MPI_Type_get_extent(basic_types[i], &lb, &basic_type_sizes[i]);
+        if (err) {
+            DTPI_Print_error(err);
+            goto fn_fail;
+        }
+        basic_type_displs[i] = displs;
+        displs += (basic_type_sizes[i] * basic_type_counts[i]);
+    }
+
+    /* create struct datatype */
+    err =
+        MPI_Type_create_struct(num_types, basic_type_counts, basic_type_displs, basic_types,
+                               &dtp->DTP_obj_array[obj_idx].DTP_obj_type);
+    if (err) {
+        DTPI_Print_error(err);
+        goto fn_fail;
+    }
+
+    err = MPI_Type_commit(&dtp->DTP_obj_array[obj_idx].DTP_obj_type);
+    if (err) {
+        DTPI_Print_error(err);
+        goto fn_fail;
+    }
+
+    dtp->DTP_obj_array[obj_idx].DTP_obj_count = 1;
+
```

Comment:
Now we only support the data with derived datatype * 1 count. Can you please add a TODO here saying that we need generate tests with derived datatype (with smaller element counts) * N counts ?